sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/batch/BulkExecutorUtil.java

```diff
@@ -94,23 +97,50 @@ static void setRetryPolicyForBulk(
 
         checkNotNull(operation, "expected non-null operation");
 
+        AtomicReference<DocumentCollection> collectionBeforeRecreation = new AtomicReference<>(null);
+
         if (operation instanceof ItemBulkOperation<?, ?>) {
             final ItemBulkOperation<?, ?> itemBulkOperation = (ItemBulkOperation<?, ?>) operation;
 
-            final Mono<String> pkRangeIdMono = BulkExecutorUtil.getCollectionInfoAsync(docClientWrapper, container)
+            final Mono<String> pkRangeIdMono = Mono.defer(() ->
+                BulkExecutorUtil.getCollectionInfoAsync(docClientWrapper, container, collectionBeforeRecreation.get())
                 .flatMap(collection -> {
                     final PartitionKeyDefinition definition = collection.getPartitionKey();
                     final PartitionKeyInternal partitionKeyInternal = getPartitionKeyInternal(operation, definition);
                     itemBulkOperation.setPartitionKeyJson(partitionKeyInternal.toJson());
 
                     return docClientWrapper.getPartitionKeyRangeCache()
                         .tryLookupAsync(null, collection.getResourceId(), null, null)
-                        .map((Utils.ValueHolder<CollectionRoutingMap> routingMap) ->
-                            routingMap.v.getRangeByEffectivePartitionKey(
+                        .map((Utils.ValueHolder<CollectionRoutingMap> routingMap) -> {
+
+                            if (routingMap.v == null) {
+                                collectionBeforeRecreation.set(collection);
+                                throw new IllegalStateException(
```

Comment:
Should we rather address the issue at the source (see https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/RxPartitionKeyRangeCache.java#L72)?
Throwing and retrying on IllegalStateException might lead to other issues later on since the exception is not specific to our implementation only. Ideally we should just leverage our CosmosException in such cases.
That said I, I'm ok with the PR as is and have a ToDo issue open to address the routingMap value being null later on.

Reply:
I think the background for the current implementation is that the RxPartitionKeyRangeCache might also be used in scenarios supporting non-partitioned collections. For bulk that is not an issue - definitely not for Spark - but for query/point operations etc. But we can certainly discuss whether there is a better approach that doesn't add regression risk.

Reply:
I think we should use a different exception type here. IllegalStateException generally means something is wrong and not recoverable. We may throw IllegalStateException from other part of SDK indicating a code bug and that may confuse the retrying logic here.

Can we change this to a subclass of CosmosException?

please see how we fixed a similar issue for the feedRange api by throwing InvalidPartitionException
https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java#L4169

Reply:
I personally think a custom exception type is overkill - but can live with it. So switched to a custom exception derived from CosmosException (using StatusCode 404 with SubStatusCode == INCORRECT_CONTAINER_RID_SUB_STATUS)